File: frontend/src/services/students.ts
/**
 * Student API services
 */
import { get, post, put } from './api';

export interface StudentProfile {
  id: number;
  user_id: number;
  full_name?: string;
  phone?: string;
  current_city?: string;
  preferred_location?: string;
  university?: string;
  degree?: string;
  graduation_year?: number;
  cgpa?: number;
  department?: string;
  year?: number;
  skills?: string;
  interests?: string;
  work_mode?: string;
  is_apaar_verified: boolean;
  apaar_id?: string;
  resume?: StudentResume;
  university_name?: string;
  start_year?: number | string;
  end_year?: number | string;
  email?: string;
}

export interface StudentResume {
  id: number;
  career_objective?: string;
  work_experience?: string; // JSON string
  projects?: string; // JSON string
  education_entries?: string; // JSON string
  skills_categorized?: string; // JSON string
  certifications?: string; // JSON string
  extra_curricular?: string; // JSON string
  resume_file_path?: string;
}

export interface Skill {
  id: number;
  name: string;
}

export interface Application {
  id: number;
  internship_id: number;
  status: string;
  applied_at: string;
}

export interface Recommendation {
  internship_id: number;
  title: string;
  company_name: string;
  match_score: number;
  matching_skills: string[];
  missing_skills: string[];
  status: string;
}

/**
 * Get student profile
 */
export const getProfile = (): Promise<StudentProfile> => {
  return get<StudentProfile>('/students/me');
};

/**
 * Update student profile
 */
export const updateProfile = (data: Partial<StudentProfile>): Promise<StudentProfile> => {
  return put<StudentProfile>('/students/me', data);
};

/**
 * Get student skills
 */
export const getSkills = (): Promise<Skill[]> => {
  return get<Skill[]>('/students/me/skills');
};

/**
 * Add a skill
 */
export const addSkill = (name: string): Promise<Skill> => {
  return post<Skill>('/students/me/skills', { name });
};

/**
 * Get my applications
 */
export const getApplications = (): Promise<Application[]> => {
  return get<Application[]>('/students/my-applications');
};

/**
 * Apply for internship
 */
export const applyForInternship = (internshipId: number): Promise<Application> => {
  return post<Application>('/students/apply', { internship_id: internshipId });
};

/**
 * Get AI recommendations
 */
export const getRecommendations = (): Promise<Recommendation[]> => {
  return get<Recommendation[]>('/students/recommendations');
};

export interface ResumeSuggestionResponse {
  suggestions: string[];
}

/**
 * Get Resume Suggestions
 */
export const getResumeSuggestions = (section: string, currentContent?: string, context?: any): Promise<ResumeSuggestionResponse> => {
  return post<ResumeSuggestionResponse>('/students/me/resume/suggestions', { section, current_content: currentContent, context });
};

export interface ResumeParseResponse {
  // Resume builder fields
  career_objective: string;
  skills: string[];
  education: Array<{
    degree: string;
    university: string;
    start_year: string;
    end_year: string;
  }>;
  experience: Array<{
    role: string;
    company: string;
    duration: string;
    description: string;
  }>;
  projects: Array<{
    title: string;
    link: string;
    description: string;
  }>;

  // Optional profile-level fields (available in /students/me/parse-resume response)
  first_name?: string;
  last_name?: string;
  email?: string;
  phone_number?: string;
  raw_text_preview?: string;
}

/**
 * Parse uploaded resume
 *
 * Backend route currently available: POST /students/me/parse-resume
 * It may return either:
 *  1) { success: true, data: {...} }  (student profile parser format), OR
 *  2) direct parser payload            (resume builder format)
 *
 * This function normalizes both into ResumeParseResponse.
 */
export const parseResume = (file: File): Promise<ResumeParseResponse> => {
  const formData = new FormData();
  formData.append('file', file);
  return post<any>('/students/me/parse-resume', formData).then((res) => {
    const payload = res?.data ?? res ?? {};
    const normalizedEducation = Array.isArray(payload.education)
      ? payload.education
      : Array.isArray(payload.education_entries)
        ? payload.education_entries.map((entry: any) => ({
            degree: entry.degree || '',
            university: entry.institution || entry.university || '',
            start_year: entry.start_year || '',
            end_year: entry.end_year || entry.year || '',
          }))
        : [];

    const normalizedExperience = Array.isArray(payload.experience)
      ? payload.experience.map((exp: any) => ({
          role: exp.role || exp.position || '',
          company: exp.company || exp.organization || '',
          duration: exp.duration || '',
          description: exp.description || '',
          start_date: exp.start_date || '',
          end_date: exp.end_date || '',
        }))
      : [];

    const normalized: ResumeParseResponse = {
      career_objective: payload.career_objective || '',
      skills: Array.isArray(payload.skills) ? payload.skills : [],
      education: normalizedEducation,
      experience: normalizedExperience,
      projects: Array.isArray(payload.projects) ? payload.projects : [],
      first_name: payload.first_name,
      last_name: payload.last_name,
      email: payload.email,
      phone_number: payload.phone_number,
      raw_text_preview: payload.raw_text_preview,
    };

    return normalized;
  });
};

/**
 * Update student resume
 */
export const updateResume = (data: Partial<StudentResume>): Promise<StudentResume> => {
  return put<StudentResume>('/students/me/resume', data);
};

/**
 * Record feedback (like, dismiss, apply)
 */
export const recordFeedback = (internshipId: number, action: string): Promise<{ status: string }> => {
  return post<{ status: string }>('/students/feedback', { internship_id: internshipId, action });
};

File: PythonProject/schemas/students.py
from pydantic import BaseModel, EmailStr, Field, field_validator
from typing import Optional, List, Dict, Any
from datetime import datetime
from .employer import InternshipOut


# 1. Signup ke waqt (Wahi purana logic)
class UserCreate(BaseModel):
    email: EmailStr
    full_name: str
    password: str
    role: str = "student"
    # Allow empty string or 12 digits
    apaar_id: Optional[str] = Field(None, description="12 digit APAAR ID (required for students)")

    @field_validator('apaar_id')
    @classmethod
    def validate_apaar(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return None
        v = v.strip()
        if not v:
            return None
        if not v.isdigit() or len(v) != 12:
             raise ValueError('APAAR ID must be exactly 12 digits')
        return v


# 2. Signup/Login Output
class UserOut(BaseModel):
    id: int
    email: EmailStr
    full_name: str
    role: str

    class Config:
        from_attributes = True


# 3. NAYA: Login ke waqt APAAR ID maangne ke liye
class LoginRequest(BaseModel):
    email: str
    password: str
    role: str  # Mandatory role check
    # Allow empty string or 12 digits
    apaar_id: Optional[str] = Field(None, description="12 digit APAAR ID")

    @field_validator('apaar_id')
    @classmethod
    def validate_apaar(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return None
        v = v.strip()
        if not v:
            return None
        if not v.isdigit() or len(v) != 12:
             raise ValueError('APAAR ID must be exactly 12 digits')
        return v



class StudentProfileUpdate(BaseModel):
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    phone_number: Optional[str] = None
    current_city: Optional[str] = None
    gender: Optional[str] = None
    languages: Optional[str] = None
    
    # APAAR ID - can be updated by existing users
    apaar_id: Optional[str] = Field(None, pattern=r"^\d{12}$", description="12 digit APAAR ID")
    
    profile_type: Optional[str] = None
    university_name: Optional[str] = None
    degree: Optional[str] = None  # e.g. B.Tech
    department: Optional[str] = None  # e.g. CSE
    year: Optional[int] = None # Keeping for backward compat, but start/end prefered
    start_year: Optional[int] = None
    end_year: Optional[int] = None
    
    cgpa: Optional[str] = None

    # Career Interests
    preferred_location: Optional[str] = None
    interests: Optional[str] = None  # e.g. Web Dev, AI
    skills: Optional[str] = None  # Comma separated skills
    projects: Optional[str] = None
    
    looking_for: Optional[str] = None
    work_mode: Optional[str] = None


# --- Resume Schemas ---

class StudentResumeBase(BaseModel):
    career_objective: Optional[str] = None
    work_experience: Optional[str] = None # JSON String
    projects: Optional[str] = None # JSON String
    certifications: Optional[str] = None # JSON String
    extra_curricular: Optional[str] = None # JSON String
    resume_file_path: Optional[str] = None
    resume_filename: Optional[str] = None
    resume_file_size: Optional[int] = None
    resume_uploaded_at: Optional[datetime] = None
    education_entries: Optional[str] = None # JSON String
    skills_categorized: Optional[str] = None # JSON String
    title: Optional[str] = None
    linkedin: Optional[str] = None
    profile_picture: Optional[str] = None

class StudentResumeCreate(StudentResumeBase):
    pass

class StudentResumeUpdate(StudentResumeBase):
    pass

class StudentResumeOut(StudentResumeBase):
    id: int
    student_id: int

    class Config:
        from_attributes = True

class ResumeParseResponse(BaseModel):
    career_objective: Optional[str] = None
    skills: List[str] = []
    education: List[Dict[str, Any]] = []
    experience: List[Dict[str, Any]] = []
    projects: List[Dict[str, Any]] = []

# 5. Profile Output (Details dikhane ke liye)
class StudentProfileOut(BaseModel):
    id: int
    user_id: int
    email: Optional[EmailStr] = None
    apaar_id: Optional[str]
    is_apaar_verified: bool
    
    first_name: Optional[str]
    last_name: Optional[str]
    full_name: Optional[str]
    
    university_name: Optional[str]
    degree: Optional[str]
    department: Optional[str]
    year: Optional[int]
    start_year: Optional[int]
    end_year: Optional[int]
    profile_type: Optional[str]
    
    cgpa: Optional[str]
    skills: Optional[str]
    interests: Optional[str]
    projects: Optional[str]
    phone_number: Optional[str]
    current_city: Optional[str]
    gender: Optional[str]
    languages: Optional[str]
    
    looking_for: Optional[str]
    work_mode: Optional[str]
    
    # Resume Data
    resume: Optional[StudentResumeOut] = None

    class Config:
        from_attributes = True


# --- Resume Schemas ---

class StudentResumeBase(BaseModel):
    career_objective: Optional[str] = None
    work_experience: Optional[str] = None # JSON String
    projects: Optional[str] = None # JSON String
    certifications: Optional[str] = None # JSON String
    extra_curricular: Optional[str] = None # JSON String
    resume_file_path: Optional[str] = None
    resume_filename: Optional[str] = None
    resume_file_size: Optional[int] = None
    resume_uploaded_at: Optional[datetime] = None
    education_entries: Optional[str] = None # JSON String
    skills_categorized: Optional[str] = None # JSON String
    title: Optional[str] = None
    linkedin: Optional[str] = None
    profile_picture: Optional[str] = None

class StudentResumeCreate(StudentResumeBase):
    pass

class StudentResumeUpdate(StudentResumeBase):
    pass

class StudentResumeOut(StudentResumeBase):
    id: int
    student_id: int

    class Config:
        from_attributes = True

class ResumeSuggestionRequest(BaseModel):
    section: str # "career_objective", "projects", "work_experience", "skills", "summary"
    current_content: Optional[str] = None # For enhancement
    context: Optional[Dict[str, str]] = None # "stream", "course", "skills"

class ResumeSuggestionResponse(BaseModel):
    suggestions: List[str]



# --- Baaki Skills aur Application wala part same rahega ---

class SkillCreate(BaseModel):
    name: str


class SkillOut(BaseModel):
    id: int
    name: str

    class Config:
        from_attributes = True


class ApplicationCreate(BaseModel):
    internship_id: int


class ApplicationOut(BaseModel):
    id: int
    internship_id: int
    status: str
    applied_at: datetime
    internship: Optional[InternshipOut] = None
    is_credit_requested: bool = False
    credit_status: Optional[str] = None
    is_pushed_to_abc: bool = False
    hours_worked: Optional[int] = None
    policy_used: Optional[str] = None
    credits_awarded: Optional[float] = None

    class Config:
        from_attributes = True


class RecommendedInternship(BaseModel):
    internship_id: int
    title: str
    company_name: str
    match_score: float
    matching_skills: List[str]
    missing_skills: List[str]
    explanation: Dict[str, Any]  # Changed from str to Dict for detailed AI matching info
    status: str = "MATCHED"
    reason: Optional[str] = None
    rule_based_score: Optional[float] = None
    embedding_score: Optional[float] = None
    feedback_boost: float = 0.0
    cross_encoder_score: Optional[float] = None

    class Config:
        from_attributes = True

class FeedbackAction(BaseModel):
    internship_id: int
    action: str # view, click, apply, ignore
